src/converter/one/convert/JfrToOtlp.java

```diff
@@ -0,0 +1,267 @@
+            int functionMark = otlpProto.startField(PROFILES_DICTIONARY_function_table);
+            int functionNameStrindex = stringPool.index(functionName);
+            otlpProto.field(FUNCTION_name_strindex, functionNameStrindex);
+            otlpProto.commitField(functionMark);
+        }
+
+        // Write location table
+        for (Line line : linePool.keys()) {
+            int locationMark = otlpProto.startField(PROFILES_DICTIONARY_location_table);
+            otlpProto.field(LOCATION_mapping_index, 0);
+
+            int lineMark = otlpProto.startField(LOCATION_line);
+            otlpProto.field(LINE_function_index, line.functionIdx);
+            otlpProto.field(LINE_line, line.lineNumber);
+            otlpProto.commitField(lineMark);
+
+            otlpProto.commitField(locationMark);
+        }
+
+        // Write string table
+        for (String s : stringPool.keys()) {
+            otlpProto.field(PROFILES_DICTIONARY_string_table, s);
+        }
+
+        // Write attributes table
+        for (KeyValue keyValue : attributesPool.keys()) {
+            int attributeMark = otlpProto.startField(PROFILES_DICTIONARY_attribute_table);
+            otlpProto.field(KEY_VALUE_key, keyValue.key);
+
+            int valueMark = otlpProto.startField(KEY_VALUE_value);
+            otlpProto.field(ANY_VALUE_string_value, keyValue.value);
+            otlpProto.commitField(valueMark);
+
+            otlpProto.commitField(attributeMark);
+        }
+
+        otlpProto.commitField(profilesDictionaryMark);
+    }
+
+    public static void convert(String input, String output, Arguments args) throws IOException {
+        JfrToOtlp converter;
+        try (JfrReader jfr = new JfrReader(input)) {
+            converter = new JfrToOtlp(jfr, args);
+            converter.convert();
+        }
+        try (FileOutputStream out = new FileOutputStream(output)) {
+            converter.dump(out);
+        }
+    }
+
+    private final class OtlpEventVisitor extends NormalizedEventVisitor {
+        private final List<SampleInfo> sampleInfos;
+
+        public OtlpEventVisitor(List<SampleInfo> sampleInfos) {
+            this.sampleInfos = sampleInfos;
+        }
+
+        @Override
+        public void visitImpl(Event event, long samples, long value) {
+            StackTrace stackTrace = jfr.stackTraces.get(event.stackTraceId);
+            if (stackTrace == null) {
+                return;
+            }
+
+            long[] methods = stackTrace.methods;
+            int[] locations = stackTrace.locations;
+
+            for (int i = 0; i < methods.length; ++i) {
+                String methodName = getMethodName(methods[i], stackTrace.types[i]);
+                int lineNumber = locations[i] >>> 16;
+                int functionIdx = functionPool.index(methodName);
+
+                Line line = new Line(functionIdx, lineNumber);
+                otlpProto.writeLong(linePool.index(line));
```

Comment:
You are writing full stack trace every time for each event. That's against the ideas behind otlp design.

Reply:
Stacktraces are already deduplicated for you in JFR constant pool.

Reply:
Right, missed that. Fixed in 885b0e9646f0f35cd0f0b73cc1875feaf48af52c